### ISSUE TITLE
Zap duration worse on 2.7.0 compared to 2.4.4

### ISSUE THREAD

**Issue by hariszukanovic** [google/ExoPlayer#3933]
### Issue description
When performing a "zap", i.e. change between two live streams, starting from v2.5.0, we experience, on average, an unexplainably longer delay before the audio/video render actually begins. 

I am not sure this is an actual bug, so I would like to start with asking for your advice in which direction to look for the cause of it...

With 2.4.4, testing different live samples, we are seeing the best case zap duration of around 0.6s and worst around 2s
With 2.7.0, also testing different live samples, we are seeing the best case zap duration of around 1s and worst around 3.5s

All of the above we reproduce with the same list of test live streams and more or less exactly the same version of our application, there is a slight difference only due to incompatibility between Exo Player version where we have added extra event listeners, nothing that should be affecting the zap performance in this way.

We perform a zap without recreating the player object, which has the playWhenReady set to true, with the following sequence:
player.stop()
player.prepare()
We see the same problem when calling both of these above with reset parameters set to true, no difference. No big difference even without calling the stop() at all.

The expected behavior for the zap time should at least to be as good, if not better, in newer versions, of course :)

In the original demo application, there is no way to perform a real zap because one needs to return to the selection activity before starting changing to a new live stream. The startup time in the demo seems very much in order, and in fact, the startup time (not zap, detail explanation below) is actually improved in 2.7.0. The zap problem is not noticeable in the demo because the stop() including whatever happens there is done on "back to stream selection activity".

So far, while trying to figure out if this was in any way bound to the way initialization is done in our app... we have discovered that there is a significant varying delay in the time when we actually receive the onAudioDisabled and onVideoDisabled events after executing the player.stop(). This also seems to be varying much more for the audio than it is for video.

In the case of 2.7.0 these events seem to be occurring after the player enters the state IDLE while in 2.4.4 they occur before. I am wondering if this can be causing the postponed start of fetching the manifest and subsequent delays of fetching the chunks as well... so that it affects worse the total zap time...

So far, in my understanding, the principal cause would be the long-lasting onAudioDisabled and onVideoDisabled. 

Filtered Logs for couple of (practically worst cases), where onAudioDisabled arrives quite late. Normally it arrives around 200ms after onVideoDisabled, but even then the total zap duration is worse using 2.7.0
Exo 2.7.0
03-03 15:21:16.383 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing ExoPlayer.STATE_IDLE duration 0.00
03-03 15:21:16.383 19402-19548/my.name.space.com I/USP_EventListeners: ZAP start[119.63]
03-03 15:21:16.411 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing ExoPlayer.STATE_BUFFERING duration 0.03
03-03 15:21:16.521 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing onVideoDisabled duration 0.14
03-03 15:21:17.773 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing onAudioDisabled duration 1.39
03-03 15:21:17.777 19402-19402/my.name.space.com D/USP_EventListeners: onLoadStarted [121.03] https://streamingserverdomainname/LIVE_1_021/cenc_manifest.mpd
03-03 15:21:18.080 19402-19402/my.name.space.com D/USP_EventListeners: onLoadCompleted [121.33] https://streamingserverdomainname/LIVE_1_021/cenc_manifest.mpd
03-03 15:21:18.090 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing onVideoEnabled duration 1.71
03-03 15:21:18.091 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing onAudioEnabled duration 1.71
03-03 15:21:18.092 19402-19402/my.name.space.com D/USP_EventListeners: onLoadStarted [121.34] https://streamingserverdomainname/LIVE_1_021/dash/cenc/init_a_eng_1520082796.m4s
03-03 15:21:18.093 19402-19402/my.name.space.com D/USP_EventListeners: onLoadStarted [121.34] https://streamingserverdomainname/LIVE_1_021/dash/cenc/init_v_r576p_1520082796.m4s
03-03 15:21:18.104 19402-19402/my.name.space.com D/USP_EventListeners: onLoadCompleted [121.35] https://streamingserverdomainname/LIVE_1_021/dash/cenc/init_v_r576p_1520082796.m4s
03-03 15:21:18.105 19402-19402/my.name.space.com D/USP_EventListeners: onLoadCompleted [121.35] https://streamingserverdomainname/LIVE_1_021/dash/cenc/init_a_eng_1520082796.m4s
03-03 15:21:18.105 19402-19402/my.name.space.com D/USP_EventListeners: onLoadStarted [121.35] https://streamingserverdomainname/LIVE_1_021/dash/cenc/chunk_a_eng_404_1520082796.m4s
03-03 15:21:18.106 19402-19402/my.name.space.com D/USP_EventListeners: onLoadStarted [121.36] https://streamingserverdomainname/LIVE_1_021/dash/cenc/chunk_v_r576p_405_1520082796.m4s
03-03 15:21:18.460 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing onVideoDecoderInitialized duration 2.08
03-03 15:21:19.382 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing onAudioDecoderInitialized duration 3.00
03-03 15:21:19.664 19402-19402/my.name.space.com D/USP_EventListeners: onLoadCompleted [122.91] https://streamingserverdomainname/LIVE_1_021/dash/cenc/chunk_a_eng_404_1520082796.m4s
03-03 15:21:19.665 19402-19402/my.name.space.com D/USP_EventListeners: onLoadCompleted [122.91] https://streamingserverdomainname/LIVE_1_021/dash/cenc/chunk_v_r576p_405_1520082796.m4s
03-03 15:21:19.666 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing onRenderedFirstFrame duration 3.28
03-03 15:21:19.678 19402-19402/my.name.space.com D/USP_EventListeners: onLoadStarted [122.93] https://streamingserverdomainname/LIVE_1_021/dash/cenc/chunk_a_eng_405_1520082796.m4s
03-03 15:21:19.678 19402-19402/my.name.space.com D/USP_EventListeners: onLoadStarted [122.93] https://streamingserverdomainname/LIVE_1_021/dash/cenc/chunk_v_r576p_406_1520082796.m4s
03-03 15:21:19.779 19402-19402/my.name.space.com D/USP_EventListeners: ZAP passing ExoPlayer.STATE_READY duration 3.40
03-03 15:21:19.780 19402-19402/my.name.space.com I/USP_EventListeners: ZAP end[123.03]
03-03 15:21:19.780 19402-19402/my.name.space.com I/USP_EventListeners: ZAP duration 3.40

Exo 2.4.4
03-03 15:15:56.439 17290-17459/my.name.space.com I/USP_EventListeners: ZAP start[78.38]
03-03 15:15:56.599 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing onVideoDisabled duration 0.16
03-03 15:15:57.629 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing onAudioDisabled duration 1.19
03-03 15:15:57.629 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing ExoPlayer.STATE_IDLE duration 1.19
03-03 15:15:57.630 17290-17290/my.name.space.com D/USP_EventListeners: onLoadStarted [79.57]https://streamingserverdomainname/LIVE_1_035/cenc_manifest.mpd
03-03 15:15:57.634 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing ExoPlayer.STATE_BUFFERING duration 1.20
03-03 15:15:57.921 17290-17290/my.name.space.com D/USP_EventListeners: onLoadCompleted [79.86]https://streamingserverdomainname/LIVE_1_035/cenc_manifest.mpd
03-03 15:15:57.932 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing onVideoEnabled duration 1.49
03-03 15:15:57.933 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing onAudioEnabled duration 1.49
03-03 15:15:57.933 17290-17290/my.name.space.com D/USP_EventListeners: onLoadStarted [79.87]https://streamingserverdomainname/LIVE_1_035/dash/cenc/init_a_eng_1519877703.m4s
03-03 15:15:57.933 17290-17290/my.name.space.com D/USP_EventListeners: onLoadStarted [79.87]https://streamingserverdomainname/LIVE_1_035/dash/cenc/init_v_r576p_1519877703.m4s
03-03 15:15:57.944 17290-17290/my.name.space.com D/USP_EventListeners: onLoadCompleted [79.88]https://streamingserverdomainname/LIVE_1_035/dash/cenc/init_a_eng_1519877703.m4s
03-03 15:15:57.944 17290-17290/my.name.space.com D/USP_EventListeners: onLoadStarted [79.88]https://streamingserverdomainname/LIVE_1_035/dash/cenc/chunk_a_eng_20872_1519877703.m4s
03-03 15:15:57.946 17290-17290/my.name.space.com D/USP_EventListeners: onLoadCompleted [79.89]https://streamingserverdomainname/LIVE_1_035/dash/cenc/init_v_r576p_1519877703.m4s
03-03 15:15:57.947 17290-17290/my.name.space.com D/USP_EventListeners: onLoadStarted [79.89]https://streamingserverdomainname/LIVE_1_035/dash/cenc/chunk_v_r576p_20884_1519877703.m4s
03-03 15:15:58.142 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing onVideoDecoderInitialized duration 1.70
03-03 15:15:58.210 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing onAudioDecoderInitialized duration 1.77
03-03 15:15:58.227 17290-17290/my.name.space.com D/USP_EventListeners: onLoadCompleted [80.17]https://streamingserverdomainname/LIVE_1_035/dash/cenc/chunk_a_eng_20872_1519877703.m4s
03-03 15:15:58.379 17290-17290/my.name.space.com D/USP_EventListeners: onLoadStarted [80.32]https://streamingserverdomainname/LIVE_1_035/dash/cenc/chunk_a_eng_20873_1519877703.m4s
03-03 15:15:58.426 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing onRenderedFirstFrame duration 1.99
03-03 15:15:58.477 17290-17290/my.name.space.com D/USP_EventListeners: ZAP passing ExoPlayer.STATE_READY duration 2.04
03-03 15:15:58.477 17290-17290/my.name.space.com I/USP_EventListeners: ZAP end[80.42]
03-03 15:15:58.477 17290-17290/my.name.space.com I/USP_EventListeners: ZAP duration 2.04

### Reproduction steps
Here I ask for directions and advice on how to debug this further.
If really needed I could try to implement this in the demo app somehow.

### Link to test content
No specific media need to reproduce the problem.

### Version of ExoPlayer being used
2.7.0

### Device(s) and version(s) of Android being used
AML S905X chipset, Android 6.0, using aml accelerated video decoder and Widevine DRM protected stream

### A full bug report captured from the device
If needed I will have to email

**Comment by ojw28** [comment]
I don't know of any changes that are likely to have made this slower. A good first step toward figuring out what happened would be to bisect the releases that occurred between 2.7.0 and 2.4.4, to work out exactly which release caused the regression. That should give a clue as to which change caused the problem. If you're able to do that and let us know, please do so!

**Comment by hariszukanovic** [comment]
We have observed a constant slower zap beginning with 2.5.0.
The 2.4.4 is the newest version that seems to work "like before". Previously we worked quite long with 2.3.1 which also did not show this behavior.

That said, I have again confirmed that also the 2.4.4 and 2.3.1 exhibit intermittent, fairly infrequent long delay (compared to average) to receive the onAudioDisabled event. Sometimes it arrives the whole 1s after the onVideoDisabled. 

On the other side, starting with 2.5.0 up to 2.7.0 the player seems to use a longer time before sending out any of the events, gradually increasing to a quite noticeable consistent delay before receiving onVideoDecoderInitialized and onAudioDecoderInitialized. 
That seems to be actually independent of delayed onAudioDisabled.

Filtered log that shows this for 2.7.0:

1. onAudioDisabled without significant delay

03-06 12:13:49.134 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onVideoDisabled duration 0.16
03-06 12:13:49.366 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onAudioDisabled duration 0.39
03-06 12:13:49.681 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onVideoEnabled duration 0.71
03-06 12:13:49.682 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onAudioEnabled duration 0.71
03-06 12:13:50.353 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onAudioDecoderInitialized duration 1.38
03-06 12:13:50.431 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onVideoDecoderInitialized duration 1.46
03-06 12:13:50.711 8916-8916/my.name.space.com I/USP_EventListeners: ZAP duration 1.74

1. onAudioDisabled delayed an extra 1s

03-06 12:13:17.215 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onVideoDisabled duration 0.17
03-06 12:13:18.444 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onAudioDisabled duration 1.40
03-06 12:13:18.784 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onVideoEnabled duration 1.74
03-06 12:13:18.785 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onAudioEnabled duration 1.74
03-06 12:13:18.987 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onVideoDecoderInitialized duration 1.94
03-06 12:13:19.576 8916-8916/my.name.space.com D/USP_EventListeners: ZAP passing onAudioDecoderInitialized duration 2.53
03-06 12:13:19.809 8916-8916/my.name.space.com I/USP_EventListeners: ZAP duration 2.76


I think these 2 are separate issues:

1. onAudioDisabled event delay is present on all tested versions, but noticeably more frequent in 2.5.0 to 2.7.0
2. onAudioDecoderInitialized event delay is consistently present 2.5.0 to 2.7.0


Not sure how much of the work of zapping gets done on the UI thread and other parallel threads... on 2.5.0 - 2.7.0... perhaps we are observing some threading prioritization issues introduced in newer versions of Exo that could explain both the frequency of 1. and consistency of 2. ?

**Comment by ojw28** [comment]
Some things you could try to help debug this:
1. Try hardcoding `MediaCodecVideoRenderer.shouldUseDummySurface` to return false. If the issue was introduced in 2.5.0, then it's possible that the changes we made around use of dummy surfaces introduced the delay.
1. Try reproducing the issue in the demo app. I don't think you need full zapping from one channel to another to measure what you're seeing, since you'll still receive `onAudioDisabled` when exiting the activity, and you should be able to time from immediately before `release` is called on the player up to that event being received.

A few additional notes:
* I think the events you're timing are being received on the main thread, so it's possible that their delivery is being delayed due to your application code doing a lot of other work on the main thread, unrelated to ExoPlayer. This is why trying to reproduce using the demo app, which does a minimal amount of work on its main thread, is a good idea.
* I tried to reproduce using `2.8.3` and the demo app, but was unable to reproduce a delay of more than ~120ms (usually <100ms).

Marking as "need more info" for now, because I don't think it's clear that this is an ExoPlayer issue. If you're able to reproduce using the demo app, please provide:
1. The small patch you're using to add timing information
1. The piece of content you're playing
1. The devices you can reproduce the problem on (ideally one we're likely to have), and to what extent it's device specific.
This will allow us to reproduce what you're seeing.

**Comment by hariszukanovic** [comment]
Thanx for pointers.
I will try to dedicate new time to this to produce some new info as soon as I can.

I am quite convinced that no other activities on main thread influence this behavior (significantly or globaly). Since I gave up on reusing the player object and I do re-instantiate it every time I do the zap, that actually gives me shorter zap time.

Just to confirm,
Is it correct to assume and expect that reusing the player object instead of re-instantiating it from scratch on streaming url change (zap) should give me shorter zap time?

**Comment by ojw28** [comment]
> Is it correct to assume and expect that reusing the player object instead of re-instantiating it from scratch on streaming url change (zap) should give me shorter zap time?

This is true as of `2.10.0`, provided you avoid calling `stop()` in-between (just call `prepare` directly with the new `MediaSource`). See [this blog post](https://medium.com/google-exoplayer/improved-decoder-reuse-in-exoplayer-ef4c6d99591d) for details.

I'm going to close this issue, since there's been no followup since August, and because `2.10.0` likely makes the zap time significantly better. If you still see larger than expected zap times, please feel free to respond here with the additional information requested in my previous response above, and we can take another look. Thanks!
